### DISCUSSION TITLE
reverse - zoom18 - with roadspeed extra tags?

### DISCUSSION THREAD

**Discussion by markdavieswlt** [osm-search/Nominatim#3432] [Q&A]
Hi Sarah, Et al

I'm using "reverse only" for a vehicle tracking use case 

My question is related to extra_tags, speed limits and buildings

Currently I use zoom=16 with extra_tags and I get the road & speed limit, this works great!

When there is a very long road, having buildings to help identify the location more accuratley on the very long road is very useful.
To do this, I can change this to zoom=18, and it will give me the nearest valid building and road, which also works great!
However, extra_tags are applied for the building and I loose the max speed for the road

Given the use case, a vehicle would be on the road, so I would want to return this max_speed of the road, but I would also want to return the nearest building, I would then format the results as
"near Standard Bank, MyVeryLong Road, London...."

Currently I have the following options
1. do 2 calls, at zoom 16 and 18 - not ideal.
2. adjust/write my own call to handle this use case

I was wondering if you had any comments on how to acheive the above, or if has been done before?
I've looked at the debug output, but thought I would ask here first before entering this rabbit hole.

Any guidance appreciated
Thanks for your time.
Mark

**Comment by lonvia** [comment]
You might want to write your own call here. The [reverse call](https://nominatim.org/release-docs/develop/library/NominatimAPI/#nominatim.api.NominatimAPI.reverse) from the library API returns quite [extensive information about the address parts](https://nominatim.org/release-docs/develop/library/Result-Handling/#detailed-address-description) of the results, including the extra tags, when address_details is set to `True`. So this means you can simply do the zoom 18 call, which gets you the housenumber point, if available, and then just look up maxspeed and friends from the extra tags of the street address.

If you don't want to write a complete new call, you can also simply change the output formatting. This is unfortunately not very well documented yet. The starting point for output formatting is in https://github.com/osm-search/Nominatim/blob/master/nominatim/api/v1/format.py. You can add another function in this file for a new custom format for the reverse call, like this:

```
@dispatch.format_func(napi.ReverseResults, 'special')
def _format_reverse_geojson(results: napi.ReverseResults,
                            options: Mapping[str, Any]) -> str:
    return my_special_format_function(results)
```

and then simply use it with the parameter `format=special`. Within the formatting function you have the full access to the result details as per library API description.

**Reply by zanetdev** [reply]
@lonvia Thanks for your reply. 
It looks like the level 18 call which returns a house number, doesn't return information (placeid / extratags) about the "parent" street object
EG:
We are trying to lookup the 'street' location with its max speed AND the interpolated "house number" address in one call to nominatim so as to not have a negative impact on performance.

![image](https://github.com/osm-search/Nominatim/assets/16229526/13113ed5-dfbf-40c2-850a-31c713ba578c)

STREET / Actual location being reverse geocoded
https://[nominatim.openstreetmap.org/reverse?lat=-34.07490&lon=18.82750&zoom=17&format=xml&addressdetails=1&extratags=1](https://nominatim.openstreetmap.org/reverse?lat=-34.07490&lon=18.82750&zoom=17&format=xml&addressdetails=1&extratags=1)

HOUSE / Interpolated location based on original street location
https://[nominatim.openstreetmap.org/reverse?lat=-34.07490&lon=18.82750&zoom=18&format=xml&addressdetails=1&extratags=1](https://nominatim.openstreetmap.org/reverse?lat=-34.07490&lon=18.82750&zoom=18&format=xml&addressdetails=1&extratags=1)

Looking at the debug view for a Level 18 Reverse Geocode
https://nominatim.openstreetmap.org/reverse?lat=-34.07490&lon=18.82750&zoom=18&format=xml&addressdetails=1&extratags=1&debug=1
![image](https://github.com/osm-search/Nominatim/assets/16229526/c7656514-e5aa-4369-8d9e-c2e2eaa4371a)
The database is queried for both street and house number information. But the street information is not actually returned. IE for a user fo the API they would have to execute the street lookup twice unless actually running against the DB itself.

As far as I can see the only way to not execute the same query twice against the DB is to actually make these calls to Postgres ourselves. Are there any other suggestions you can provide other than us doing this?

Would I be correct in saying that in the formatting of the results, we do not have access to anything other than what is returned in the usual api response payload? (IE it won't have the tags for the street if we are querying at level 18 and it returns a house number)

**Reply by lonvia** [reply]
> Would I be correct in saying that in the formatting of the results, we do not have access to anything other than what is returned in the usual api response payload?

That is not correct. The datastructure used to format the responses has a lot more information. Please have a look at the documentation links I posted above.

**Reply by zanetdev** [reply]
> > Would I be correct in saying that in the formatting of the results, we do not have access to anything other than what is returned in the usual api response payload?
> 
> That is not correct. The datastructure used to format the responses has a lot more information. Please have a look at the documentation links I posted above.

Thank you Sarah, the formatting of the output result has solved our issue. 
For others, new documentation available here : https://nominatim.org/release-docs/latest/customize/Result-Formatting/
